appliance/postgresql2/state/state.go

```diff
@@ -0,0 +1,1019 @@
+		log.Info("configured for normal mode but found cluster in singleton mode, transitioning cluster to normal mode", "at", "transition_singleton")
+		if p.clusterState.Primary.ID != p.self.ID {
+			panic(fmt.Sprintf("unexpected cluster state, we should be the primary, but %s is", p.clusterState.Primary.ID))
+		}
+		p.startTransitionToNormalMode()
+		return
+	}
+
+	// The primary peer needs to check not just for liveness of the synchronous
+	// peer, but also for other new or removed peers. We only do this in normal
+	// mode, not one-node-write mode.
+	if p.singleton {
+		return
+	}
+
+	// TODO: It would be nice to process the async peers showing up and
+	// disappearing as part of the same cluster state change update as our
+	// takeover attempt. As long as we're not, though, we must handle the case
+	// that we go to start a takeover, but we cannot proceed because there are
+	// no asyncs. In that case, we want to go ahead and process the asyncs, then
+	// consider a takeover the next time around. If we update this to handle
+	// both operations at once, we can get rid of the goofy boolean returned by
+	// startTakeover.
+	if !p.peerIsPresent(p.clusterState.Sync) &&
+		p.startTakeover("sync gone", p.clusterState.InitWAL) {
+		return
+	}
+
+	presentPeers := make(map[string]struct{}, len(p.clusterPeers))
+	presentPeers[p.clusterState.Primary.ID] = struct{}{}
+	presentPeers[p.clusterState.Sync.ID] = struct{}{}
+
+	newAsync := make([]*discoverd.Instance, 0, len(p.clusterPeers))
+	changes := false
+
+	for _, a := range p.clusterState.Async {
+		if p.peerIsPresent(a) {
+			presentPeers[a.ID] = struct{}{}
+			newAsync = append(newAsync, a)
+		} else {
+			log.Debug("peer missing", "at", "missing_async", "async.id", a.ID, "async.addr", a.Addr)
+			changes = true
+		}
+	}
+
+	// Deposed peers should not be assigned as asyncs
+	for _, d := range p.clusterState.Deposed {
+		presentPeers[d.ID] = struct{}{}
+	}
+
+	for _, peer := range p.clusterPeers {
+		if _, ok := presentPeers[peer.ID]; ok {
+			continue
+		}
+		log.Debug("new peer", "at", "add_async", "async.id", peer.ID, "async.addr", peer.Addr)
+		newAsync = append(newAsync, peer)
+		changes = true
+	}
+
+	if !changes {
+		return
+	}
+
+	p.startUpdateAsyncs(newAsync)
+}
+
+func (p *Peer) startInitialSetup() {
+	if p.updatingState != nil {
+		panic("already have updating state")
+	}
+
+	p.updatingState = &State{
+		Generation: 1,
+		Primary:    p.self,
+		InitWAL:    xlog.Zero,
+	}
+	if p.singleton {
+		p.updatingState.Singleton = true
+		p.updatingState.Freeze = NewFreezeDetails("cluster started in singleton mode")
+	} else {
+		p.updatingState.Sync = p.clusterPeers[1]
+		p.updatingState.Async = p.clusterPeers[2:]
+	}
+	log := p.log.New("fn", "startInitialSetup")
+	log.Info("creating initial cluster state", "generation", 1, "at", "create_state")
+
+	if err := p.putClusterState(); err != nil {
+		log.Error("failed to create cluster state", "at", "create_state", "err", err)
```

Comment:
Yeah, looks like this error path needs some cleanup, I'll figure out what it should do (probably schedule another evaluation after a second).